Add: two posts, one about our infrastructure, and one about our recent migration to it (#267)
OpenTTD/website: _posts/2023-07-09-infra-migration.md

---
title: The 2023 infrastructure migration
author: TrueBrain
---

This post is part of a two-part series, of which this is the second.
In the [first post](/news/2023/07/08/our-infrastructure), we went into a bit into the infrastructure that runs OpenTTD, from BaNaNaS to our main website.
In this post, we will explain a bit about the migration we just did to get to this infrastructure.

After over 2 months of work, I am happy to announce we finished ([another](https://www.openttd.org/news/2020/12/01/infrastructure-migration-finished)) infrastructure migration.
Today is the day I removed the last few DNS entries pointing to [AWS](https://aws.amazon.com/)'s DNS servers, and I am proud to mention that (almost) all traffic is now routed via [Cloudflare](https://cloudflare.com/) (and we aren't even receiving sponsoring to say so).

In this post I want to take you with me why this migration was needed, what the benefits are, and why you possibly care.
But in short summary:

- (Much) smaller monthly bill as AWS charges insane amounts for bandwidth.

> - Faster download speeds for you (ranging from the in-game content service to download the game from our website).

Reviewer: ```suggestion
- Faster download speeds for you (ranging from the in-game content service to downloading the game from our website).
```

Back then, AWS was offering Open Source credits, which meant we didn't have to worry so much about money.
Additionally, they had everything we needed, and it seemed like a good fit.

This has been a very good call; not only did I personally have far less worries about things like security, stability, etc, we also had the fewest interruption till date.
More than 99.99% uptime over all our services since then.
This means it has been very unlikely you couldn't download the game, or use the in-game content service, etc.
Even big companies would be very happy with 99.99%, but given we run this with only a handful of people in their free time ... most excellent, if you ask me.

The fun thing is, with such uptime, when we got a report of a problem: it is very likely to be a user-problem, and not a backend-problem.
This is a great spot to be in!

## Problems on the horizon

But, for more than a year now, there have been some issues that were hard to deal with.
Let's explain a few in a bit more detail.

### AWS's bandwidth price

The Open Source credits ran out after the first year; and where we hoped (and somewhat assumed) the AWS bandwidth prices would go down over time, they did not.

> For those who don't know, AWS charges 9 dollar cents per GB of bandwidth.

Reviewer 2: Is this a common way to specify US Dollar cents? I've never seen it (as an American where all cents are assumed to be USD).

All your other dollar amounts use the world dollar, so I don't know if you'd want to change it, but I generally write money with the $ sign, so $0.09.

Dunno what to do here, feel free to leave it as-is. 🤷 😄

Author: changed them all in USD.

If AWS Cloudformation templates work, they work.
But if they don't, they really do not.
This mostly is a problem when renaming objects, or moving them around a bit.
And guess what you often had to do when you tried to update to a newer version of AWS's CDK?
You guessed right!

In short, this has been the worst decision we could have made back in 2020.
I still don't actually regret it, but I would not advise anyone to use AWS's CDK ever.
(and if you want to know more, as there are many more reasons, feel free to send me an email or a [Discord](https://discord.gg/openttd) DM; I am happy to rant some more.)

### The bill

OpenTTD lives on donations.
We haven't had a fundraiser in years, mainly because at a certain moment we had more than 5000 pounds in reserve.
This meant we also weren't that scared that AWS would cost a bit more than dedicated servers.
If it would mean we all had to work less to keep things running, and that it would be more secure too, that would be worth the money.

But, the cost of AWS did mean the funds started to go down.
We yearly got less donations than the costs.

An average month on AWS currently costs us 280 dollars (so yearly roughly 3500 dollars).

> Of course this is peanuts for most people, but for an Open Source project, that is a lot.

Reviewer 2: ```suggestion
Of course this is peanuts for most companies, but for an Open Source project, that is a lot.
```
3500 dollars isn't exactly peanuts for me, maybe you meant companies. 😉

Terraform has been around longer, and although they do offer a CDK now, where you can use Python to write Terraform (instead of their custom HCL), it still isn't really there yet.
But then there is Pulumi, which offers exactly what we need.
And, what is always nice, they also offer Open Source [a free account](https://github.com/pulumi/team-edition-for-open-source/issues/14) to use their Pulumi Cloud.

So, we went for Pulumi.
And I have enjoyed the experience very much.

The most beautiful moment was when I read they had a new version released .. and of course I was: meh, this will take me a while.

Nope.

Two minutes.
All done.
That just makes you cry of happiness.

In the background most of Pulumi's registries use Terraform, so there is a lot of reuse of existing known-good work.
And this makes me pretty sure we made a solid choice here.

Pulumi allows us to provision infrastructure for:
- AWS
- Nomad (more on that in a bit)
- Cloudflare
- GitHub

and possibly anything else we will find in the next few years.
This means that one place, our [infrastructure repository](https://github.com/OpenTTD/infra), defines everything infra-related for all parts it touches.

> Our infrastructure isn't small: 3500 LoC to define it.

Reviewer 2: ```suggestion
Our infrastructure isn't small: 3500 lines of code to define it.
```
No need for an acronym here 😉

And the metrics I actually want, would cost so much money a month, that I didn't find that reasonable.

So after looking around a bit, I quickly found [Nomad](https://www.nomadproject.io/), from Hashicorp.
It basically does the same as AWS ECS, but you can host it on what-ever you want.
As you might have guess: that is what we are running as our backend now, and all containers are running happily.

Using Nomad hasn't been a flawless experience.
As we want to be as cheap as possible (which happens when you depend on donations), we are using AWS's T4g EC2-instances.
These machines are ARM64.
And Nomad had some open issues in those regards.

But, nothing that can't be fixed, and so far they have been great accepting contributions.
That is the lovely thing about a company that makes their products Open Source: others can contribute to make it better.
So by now I fixed the two problems that were holding me back from using it on ARM64, and I hope others enjoy those fixes too.

With Nomad, I have full insight what a service is doing.
How much memory, how much CPU, how often they crashed, etc.
The one small thing I am still missing: I cannot see which service uses how much bandwidth.

> But I plan on addressing that soon .. I have some ideas.

Reviewer 3: ```suggestion
But I plan on addressing that soon... I have some ideas.
```

That way, new machines are created with the latest version, and the old ones are slowly shut down.
Far far easier than trying to manually update machines; and much less likely to break.

## GitHub Workflows

The one thing not talked about above, is that we also really badly needed to address our [GitHub Workflows](https://docs.github.com/en/actions/using-workflows).
GitHub Workflows automate our repositories, but were written in a time where [Reusing Workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) didn't exist yet.

The problem was maintaining the GitHub Workflows; it would require changes in ~20 repositories.
This was very time consuming and error-prone.

By switching to Reusing Workflows, it means we have a [single repository](https://github.com/OpenTTD/actions) which hosts all our workflows.
By changing things in a single place, all repositories will be updated.

This will make our life so much easier.

While at it, we also noticed that making releases of those repositories to get things to production wasn't really working, and was an annoying task.

> So we decided to also change that: if you now merge a Pull Request, it is deployed to production.

Reviewer 2: ```suggestion
So we decided to also change that: if we now merge a Pull Request, it is deployed to production.
```
The phrasing here sounded off to me. Only "we" merge Pull Requests and add Preview labels, not most people reading this ("you").

Same for a couple places below.

No delay.

> And to compensate for the missing staging, you can now label a Pull Request with a preview label in any of the repositories, and it will deploy this to a preview site.
> Here you can review the changes, and if you like it, merge the Pull Request.

Reviewer 2: ```suggestion
And to compensate for the missing staging, we can now label a Pull Request with a preview label in any of the repositories, and it will deploy this to a preview site.
Here we can review the changes, and if we like it, merge the Pull Request.
```

And boom, it is in production.

## Migration finished; what did we gain?

Now after a few weeks, well, months of work, I can finally say we migrated everything over.
And I am, so far at least, very happy with the result.

The AWS bill is reduced with at least 50%.
Cloudflare does cost us a bit of money: 20 dollar a month for Pro-plan, 5 dollar a month for Cloudflare Workers, and like 2 dollar a month for Cloudflare R2.
But overall the total bill should be reduced with at least 100 dollar a month (so 1/3rd), but most likely more.
We will have to wait another month to get a better picture of the bill.

We also applied for the Cloudflare Open Source, but haven't heard back from them yet.
Maybe they have forgotten us.
It would reduce our bill with the above mentioned 27 dollar per month.
But honestly, even if they did forget about us, the amount of money Cloudflare costs versus AWS: it is totally worth paying for it.
The contrast between these two companies really is day and night, in many ways.
Not only about how they charge customers, and what they charge them for.
But also [Cloudflare's documentation](https://developers.cloudflare.com/) is one of the best I know.

> Further more, over 50% of the BaNaNaS content is cached by Cloudflare.

Author: They were never hosted separately :) So yeah, it is within that number :)

So if you are living in America or China, it is very likely downloads are now a lot faster.

Over 60% of the content on our CDN, like releases etc, are cached by Cloudflare.
So there too, it is very likely that if you download the latest release, it goes a lot faster.

And, what I personally enjoy, I have full insight of this data.
I can now tell we consume about 1.3TB of bandwidth a week, and that we have ~57k unique visitors a week.
There are 2.5M request made in the last 7 days.
Knowing this information means we can better anticipate what changes we need to make, and if we change something, what the impact would be.

But, we only finished the migration today; so we will need some more time to say if this was actually a good move.
It sure feels like one.

## In closing

Although we only migrated our infrastructure only 2.5 years ago, this migration was really needed.
And I really hope we don't need another migration in 2.5 years; at least I did my best to set it up as such that it wouldn't be needed.

> As for the donations, we are all still surprised how generous you all our with your donations without us asking for it.

Reviewer 3: ```suggestion
As for the donations, we are all still surprised how generous you all with your donations without us asking for it.
```
